### DISCUSSION TITLE
How to get the syntax node, given a location

### DISCUSSION THREAD

**Discussion by martinrhan** [dotnet/roslyn#53746] [Q&A] [answered]
I am currently learning roslyn source generator. What I intend to make is a source generator that automatically implements interfaces property member when a type interit interface but corresponding members are not implemented. The automaticlly implemented members will be autoproperty by default.

What I have done so far:
```
    [Generator]
    public class AutoImplementInterfaceGenerator : ISourceGenerator {
        public void Initialize(GeneratorInitializationContext context) {
            context.RegisterForSyntaxNotifications(() => new MySyntaxContextReceiver());
            
        }

        public void Execute(GeneratorExecutionContext context) {
            List<Location> Locations = new List<Location>();
            foreach (Diagnostic diagnostic in context.Compilation.GetDeclarationDiagnostics()) {
                if (diagnostic.Id == "CS0535" && !Locations.Contains(diagnostic.Location)) {
                    Locations.Add(diagnostic.Location);
                }
            }
            foreach (Location location in Locations) {
                SymbolFinder.FindSymbolAtPositionAsync()
            }
        }
    }

```
As you can see I am stuck on finding the `SyntaxNode` of where [Compiler Error CS0535][1] occur, which is represented by `Diagnostic.Location`(See [here][2]). An approach I found was through `SymbolFinder.FindSymbolAtPositionAsync`(See [here][3]) method, however it requires a `Document`or `Workspace`. I have no idea about how to get a instance of that within APIs related to source generator.

  [1]: https://docs.microsoft.com/en-us/dotnet/csharp/misc/cs0535?f1url=%3FappId%3Droslyn%26k%3Dk(CS0535)
  [2]: https://docs.microsoft.com/en-us/dotnet/api/microsoft.codeanalysis.location?view=roslyn-dotnet-3.9.0
  [3]: https://docs.microsoft.com/en-us/dotnet/api/microsoft.codeanalysis.findsymbols.symbolfinder.findsymbolatpositionasync?view=roslyn-dotnet-3.9.0#Microsoft_CodeAnalysis_FindSymbols_SymbolFinder_FindSymbolAtPositionAsync_Microsoft_CodeAnalysis_Document_System_Int32_System_Threading_CancellationToken_

**Comment by jmarolf** [comment, marked as answer]
the location type has a reference to the syntax tree the location comes from. You need to get the "Root" node (the first node in the tree) and then call FindNode on the root to get the nearest child node at the given span. Should look something like this:

```csharp
foreach (Location location in Locations) {
    SyntaxNode rootNode = location.SourceTree.GetRoot();
    SyntaxNode nodeAtLocation = rootNode.FindNode(location.SourceSpan);
}
````
